### DISCUSSION TITLE
Add Progressive Web App Service worker

### DISCUSSION THREAD

**Discussion by chsasank** [outline/outline#1337] [General]
Outline ticks most of the checkboxes required for PWA. See here: https://www.pwabuilder.com

<img width="1522" alt="image" src="https://user-images.githubusercontent.com/9305875/86736726-605fda80-c051-11ea-962a-15254b42325c.png">

We just need to add service worker. Can we add the pwa service worker and make outline complete progressive web app? We just need to follow instructions here: https://www.pwabuilder.com/serviceworker. It's a fairly simple process of adding an extra script to `<head>` and a static js file.

**Comment by tommoor** [comment]
You could "tick the box" by adding a no-op service worker for sure, but I think that would be somewhat defeating the object of the PWA?

 If you had a service worker that did no caching would that still have advantages over none at all? I'm actually not sure.

**Comment by chsasank** [comment]
Yeah agree that we should use caching service worker.

For example, we can use code snippet for `Offline copy with Backup offline page`. If you can guide me through where I should modify the files, I can make this PR by end of this week.

**Reply by tommoor** [reply]
I would imagine such a file would live in the public directory next to the manifest?

https://github.com/outline/outline/blob/develop/public/manifest.json

**Reply by chsasank** [reply]
Sorry for asking noob stuff, what about `<header>`? Which file do I edit for that. There are so many headers when I search: https://github.com/outline/outline/search?p=1&q=header&unscoped_q=header.

Thanks.

**Reply by tommoor** [reply]
Here: https://github.com/outline/outline/blob/5010b08e83bc98701bee30734fcf8a95ccd1666e/server/static/index.html#L8

It would [be here](https://github.com/outline/outline/blob/f43deb7940e01eeb7c4f3ec11c6aa505e39bd747/server/pages/components/Layout.js#L48), but there [is a PR](https://github.com/outline/outline/pull/1331) in that will get merged in the next 48 hours that refactors server side rendering and removes this file.

**Reply by chsasank** [reply]
Thanks a lot for help and headsup :).

> but there is a PR in that will get merged in the next 48 hours

Oh well, then I'll work on this on weekend after that PR's merged.

**Comment by almereyda** [comment]
I see possible good use coming from having an outline wiki accessible offline. Now that #811 is due with #1660 , there is also a data structure that allows bringing in content changes from the time during a network split. Is there a chance we want to put the application in a service worker, in so it becomes available offline?

https://github.com/fiduswriter/fiduswriter have recently shown how to use a service worker to allow offline editing of text content, and merging it back with the online version once reconnected.

**Comment by philipgierszal** [comment]
Hey,
any update on this? I would love to help, desktop PWA is very much needed!

**Reply by tommoor** [reply]
This has been available for some time

**Reply by philipgierszal** [reply]
Thank you for your swift reply! I followed [this ](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/from-source-BlBxrNzMIP) to set up a self hosted version. Unfortunately while inspecting through Google Chrome 98.0.4758.82, going to the Application section, i see this: 
![service_worker](https://user-images.githubusercontent.com/49032619/154034788-e8bcba7b-df1d-489e-ad57-a062483558dd.PNG)
Am I doing something wrong?

**Reply by philipgierszal** [reply]
Sorry, turns out it was my Apache2 config, just had to add the line:

`Header set Service-Worker-Allowed "/"`
